doc/design/publish-1.md

```diff
@@ -0,0 +1,141 @@
+Published Versions are immutable, only draft Versions are editable (except admins, who have superpowers).
+Versions also track their validation status.
+
+### AssetBlob
+This is basically just a pointer to a blob in S3.
+AssetBlobs are central to how we implement deduplication of blobs in S3.
+Multiple Assets can refer to the same AssetBlob, meaning we only need to store each blob in S3 once.
+
+Whenever a new AssetBlob is created, a background task runs to compute the sha256 checksum (a "digest") of the blob.
+This can take a long time for large blobs, as it must stream all of the data from S3.
+
+### Asset
+Assets consist of a JSON blob of metadata and a reference to an AssetBlob.
+Versions and Assets are many-to-many; a Version can have any number of Assets, and an Asset can belong to any number of Versions (even, hypothetically, versions in different Dandisets).
+Assets also track their validation status.
+
+Similarly to Versions, there are two kinds of Asset: draft and published.
+Draft Assets are only allowed to belong to draft Versions.
+Published Assets are basically the same as draft Assets, but have some extra metadata included that was defined during the publish process (`datePublished`, `publishedBy`, etc).
+
+Note that published Assets can belong to both draft and published Versions.
+
+Uploading new data or modifying the metadata of an Asset doesn't actually modify the Asset.
+Instead, it creates a new Asset and replaces the reference to that Asset in the Version.
+This ensures that other Versions which might refer to the original Asset are unmodified, while the Version containing the Asset being modified reflects the changes correctly.
+
+When a published Asset is modified in this way, the publish metadata is stripped out, effectively creating a new draft Asset.
+
+_In general_, Assets should be immutable; any modification will instead create a new draft Asset.
+* One exception to this is calculating the AssetBlob digest.
+When the digest completes, it will be injected directly into the metadata of any Assets referencing that AssetBlob.
+* The other exception is creating a published Asset during the publish process.
+This involves directly injecting the new metadata fields into the draft Asset, effectively turning it into a published Asset in place.
+
+### Metadata Schemas
+Version metadata and Asset metadata must conform to defined schemas.
+There are two sets of schemas.
+The first applies to both draft and published metadata, and is used by the web metadata editor.
+The second is more restrictive, and only applies to published metadata.
+It requires additional fields like `datePublished`, `publishedBy`, and `doi` that are only knowable at publish time.
+The second one is used to validate metadata prior to publish.
+
+The JSON schemas can be found at https://github.com/dandi/schema.
+
+The pydantic models can be found and imported from https://github.com/dandi/dandischema.
+dandi-api will use the pydantic models for validation.
```

Comment:
At least ATM we have no provisioning for validating against different schema versions via pydantic and I do not see yet how we could in the long(er) run due to all possible dependencies on pydantic version/logic etc.  Multiple strategies could be explored (e.g. "upgrade provided metadata to the most recent pydantic model version and validate") but we do not yet have support for that even designed.  Filed https://github.com/dandi/dandischema/issues/5 .

In the short term, I guess we can go "pydantic model" way but would need to orchestrate reextraction (for assets) and upgrade (for dandisets) of all metadata records and make publish accept only the most recent one. Thus here we would need to abstract away IMHO into

```suggestion
dandi-api will use the `dandischema` validation API (ATM will not support older schema versions and thus require metadata re-extraction/upgrade).
```
or smth like that.

Reply:
My understanding was that it was feasible to write a library that would return the correct pydantic models for a given version. 

```suggestion
dandi-api will look up the JSON schema for the specified `schemaVersion` and use it for validation. Hopefully at some point we can find a way to provide arbitrary pydantic model versions and use those instead.
```